osmnx/simplify.py

```diff
@@ -354,4 +371,160 @@ def clean_intersections(G, tolerance=15, dead_ends=False):
     # get the centroids of the merged intersection polygons
     unified_intersections = gpd.GeoSeries(list(buffered_nodes))
     intersection_centroids = unified_intersections.centroid
-    return intersection_centroids
+
+    # To make things simpler, every edge should have a geometry
+    # (this avoids KeyError later when choosing between several
+    #  geometries or when the centroid inherits a geometry directly)
+    for u, v, data in G.edges(keys=False, data=True):
+        if 'geometry' not in data:
+            # if it doesn't have a geometry attribute, the edge is a straight
+            # line from node to node
+            x1 = G.nodes[u]['x']
+            y1 = G.nodes[u]['y']
+            x2 = G.nodes[v]['x']
+            y2 = G.nodes[v]['y']
+            data['geometry'] = LineString([(x1, y1), (x2, y2)])
+
+    # First, we need to find the nearest centroid to each node
+    if cKDTree:
+        # the search is faster if we use a cKDTree
+        # build a k-d tree for euclidean nearest node search
+        centroids = pd.DataFrame({'x': list(map(lambda c: c.x, intersection_centroids)),
+                                  'y': list(map(lambda c: c.y, intersection_centroids))})
+
+        tree = cKDTree(data=centroids[['x', 'y']], compact_nodes=True, balanced_tree=True)
```

Comment:
We could generalize the method `get_nearest_nodes()` to take in as first argument the points used to form the kd-tree. But that would mean changing every existing call to `get_nearest_nodes` to pass the graph's nodes as input instead of `G`:

```python
nodes = pd.DataFrame({'x':nx.get_node_attributes(G, 'x'),
                      'y':nx.get_node_attributes(G, 'y')})
```
which is a bit of a hassle.

Ideally, we would overload `get_node_attributes()` to keep a single signature, whilst supporting different input types. As overloading is not really a thing in python, we can either:

1. Add a second parameter `points` that defaults to None . Then, either `G` or `points` could be passed by the user, but not both. However, I think this can be a bit confusing and not very elegant.
2. Have a generic method that takes points as input and does nn search, and modify `get_nearest_nodes` internally to call the first method. 

For instance (option 2):

```suggestion
def nearest_neighbors_k2tree(nodes, X, Y):
    """
    Return the nearest neighbors to a list of points, in a 2-dimensional
    space indexed by nodes.
    
    Parameters
    ----------
    nodes : pandas DataFrame with columns 'x' and 'y'
        The data points of dimension 2 to be indexed

    points : ndarray, shape (n,2)
        The n data points of dimension 2 for which we want to find the
        nearest neighbors.

    Returns
    -------
    nn : array
        list of nearest neighbors
    """

    start_time = time.time()

    if not cKDTree:
        raise ImportError('The scipy package must be installed to use this optional feature.')

    # build a k-d tree for euclidean nearest node search
    tree = cKDTree(data=nodes[['x', 'y']], compact_nodes=True, balanced_tree=True)

    # query the tree for nearest node to each point 
    points = np.array([X,Y]).T   
    dist, idx = tree.query(points, k=1)
    nn = nodes.iloc[idx].index

    log('Found nearest nodes to {:,} points in {:,.2f} seconds'.format(len(nodes), time.time()-start_time))

    return np.array(nn)
    
def get_nearest_nodes(G, X, Y, method=None):
...
 
    elif method == 'kdtree':
        
        nodes = pd.DataFrame({'x':nx.get_node_attributes(G, 'x'),
                              'y':nx.get_node_attributes(G, 'y')})
        nn = nearest_neighbors_k2tree(nodes,X,Y)
...   
```

We could do the same for the other methods and even generalise further to k dimensions. But I'm not sure that we gain a lot by this kind of generalisation, especially if it's not used very often. We're effectively building a very thin wrapper around the scipy API.

Reply:
@gboeing I might have found an alternative that reuses `ox.get_nearest_nodes` without the need to modify it. We first create the new graph, add the nodes, and then call `ox.get_nearest_nodes` on that new graph:

```python
# Add the nodes first
# The centroids are given new ids = 0 .. total_centroids-1
for centroid, elements in centroid_elements.items():
    x = intersection_centroids[centroid].x
    y = intersection_centroids[centroid].y

    # we then have to do this later as we don't have the mapping node -> centroid yet
    # osmid = elements[0] if len(elements) == 1 else elements
    G__.add_node(centroid, x = x, y = y)

# get the nodes of G as points
X = [ x for node, x in G.nodes(data = 'x') ]
Y = [ y for node, y in G.nodes(data = 'y') ]
points = np.array([X, Y]).T

nearest_centroids = ox.get_nearest_nodes(G__, points)

# now we have the mapping node -> centroid
```

Reply:
@gboeing shall I go ahead with this improvement?

Reply:
@PedrosWits I think this looks like a good idea. See also #261 for some related discussion regarding finding nearest edges. cc @juliamgomes

Reply:
addressed in [b4caee6](https://github.com/gboeing/osmnx/pull/249/commits/b4caee6902410aa331a926dfac1256f2d330447f)